### PR TITLE
Fixes for NodeJS.EventEmmiter and for socket.io.Socket. Also add socket.io.users types

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by kataras** [comment]
Only this error if I add the correct _events (you should use it in order to get a list of registed events,it wasn't there I added it) property on nodejs.eventemitter CLASS.
----------------- For file:pty.js/pty.js.d.ts
pty.js/pty.js.d.ts(23,18): error TS2420: Class 'Terminal' incorrectly implements interface 'Stream'.
  Types of property 'addListener' are incompatible.
    Type '(event: string, listener: Function) => EventEmitter' is not assignable to type '(event: string, listener: Function) => EventEmitter'.
      Type 'NodeJS.EventEmitter' is not assignable to type 'EventEmitter'.
        Property '_events' is optional in type 'EventEmitter' but required in type 'EventEmitter'.

File: node/node.d.ts [hunk 1]
```diff
@@ -170,6 +170,7 @@ declare module NodeJS {
         removeAllListeners(event?: string): EventEmitter;
         setMaxListeners(n: number): void;
         listeners(event: string): Function[];
+        _events?:string[];
```
Comment (basarat) [inline node/node.d.ts]: Agreed that it does indeed exist ... however we do not support *public* access to internal implementations so this cannot be merged. 

Docs : https://nodejs.org/api/events.html

**Comment by basarat** [comment]
Two questions: 

* why do we want `_events` exposed. An *example* of a use case would be appriciated.
* what is going on with `_rooms` vs. `rooms`. The comments seem to imply a socket.io bug? A reference. 

Thanks for the PR :rose:

**Comment by kataras** [comment]
1) for me 'listeners'  doesn't return the list, but _events (this is the property I knew for collection) is working. Look at: https://github.com/nodets/socket.io.users/blob/master/src/lib/Users.ts#L147 

2) rooms didn't working also, I use ._rooms and working. Look at: https://github.com/nodets/socket.io.users/blob/master/src/lib/User.ts#L141

on 1) the listeners method excepting an event:string parameter, the deftyped of node.d.ts doesn't have a property to return all registed events.

on 2) the socket.io.d.ts has a property named 'rooms' but this doesn't return the rooms that the socket is joined to, the correct property is _rooms.

I had a project this project written in javascript with ._events and ._rooms, I converted it to typescript and I saw , listeners and .rooms, I tested it and didn't working I spend one day to understand what was the wrong....

( I Know  these because I read all source of eventemitter and socket.io)

Thanks for replying !

File: node/node.d.ts (hunk 1 above)
Comment (kataras) [inline node/node.d.ts]: But how can I get all events that are registed  to with the existing node.d.ts ? Sometimes you need all events no just the listeners(eventType)

**Comment by basarat** [comment]
> but _events (this is the property I knew for collection) is working

*nit events is not a `string[]`. More like `any` : https://github.com/nodejs/node/blob/abb2a4b8fe545964979525a330b9533d9a8cfbd5/lib/events.js#L34

But again since its not documented I recommend using `['_events']` in your local code. This change to `node.d.ts` will not be approved. 


> The socket.io.d.ts has a property named 'rooms' but this doesn't return the rooms that the socket is joined to, the correct property is _rooms.

Please report as a bug to https://github.com/socketio/socket.io/issues. Again an internal API will not be exposed without official documentation.

^ I know this sounds bad and I do appreciate your effort ... but these are just the standards that are followed for this repo.

Other than that we'd be happy to merge `socket.io.users` definition + tests. :rose:

**Comment by basarat** [comment]
> but these are just the standards that are followed for this repo.

I've just added this to the contributing guide as well: http://definitelytyped.org/guides/contributing.html

File: node/node.d.ts (hunk 1 above)
Comment (basarat) [inline node/node.d.ts]: Use `['_events']` if you want to. This makes it explicit that you are using undocumented code paths

**Comment by kataras** [comment]
Oh I understand and you have right ... although I though I could help other developers that may see this anomaly on their code too, but no problem we will wait for official response.

Yes please merge socket.io.users and at 2-3 hours node-mysql-wrapper new d.ts  is coming  :P Thanks for answering!!

**Comment by basarat** [comment]
> Yes please merge socket.io.users

Will do. Either open a new PR with just those changes or revert the changes to the other files. In either case ping me once done :rose:  Thanks!

**Pull request closed without merging**